src/thread_uv_iface.c

```diff
@@ -0,0 +1,270 @@
+    jl_init_thread_heap(ptls);
+    jl_install_thread_signal_handler(ptls);
+
+    jl_all_tls_states[tid] = ptls;
+    jl_tp_init_done = 1;
+}
+
+// struct queued and executed on worker threads (via the UV API).
+typedef struct {
+    uv_work_t       req;           // libuv reference
+    tp_workfun_t    workfun;       // Julia function run on worker thread
+    void            *workfun_arg;  // arg to workfun (actually a jl_value_t*)
+
+    tp_notify_cb_t  notify_cb;     // libuv calls this when workfun completes in the main thread
+    int             notify_ref;    // arg to notify_cb
+
+    jl_module_t     *current_module;
+    size_t          world_age;
+} tp_work_t;
+
+
+// libuv calls this on a worker thread
+void tp_run_work(uv_work_t *req)
+{
+    if (!jl_tp_init_done)
+        tp_initthread();
+
+    // Stuff copied over from the current thread implemention.
+    // TODO : cleanup
+    jl_ptls_t ptls = jl_get_ptls_states();
+    jl_init_stack_limits(0);
+
+    // set up tasking
+    jl_init_root_task(ptls->stack_lo, ptls->stack_hi - ptls->stack_lo);
+#ifdef COPY_STACKS
+//    jl_set_base_ctx((char*)&arg);
+#endif
+   jl_gc_state_set(ptls, JL_GC_STATE_SAFE, 0);
+
+
+    tp_work_t *p_work = (tp_work_t *) req->data;
+
+    int8_t gc_state = jl_gc_unsafe_enter(ptls);
+    // This is probably always NULL for now
+    jl_module_t *last_m = ptls->current_module;
+    size_t last_age = ptls->world_age;
+    JL_GC_PUSH1(&last_m);
+    ptls->current_module = p_work->current_module;
+    ptls->world_age = p_work->world_age;
+
+    //printf("Running workfun %p %p\n", p_work->workfun, p_work->workfun_arg);
+    p_work->workfun(p_work->workfun_arg);
+
+    ptls->current_module = last_m;
+    ptls->world_age = last_age;
+    JL_GC_POP();
+    jl_gc_unsafe_leave(ptls, gc_state);
+}
+
+// libuv calls this in the main thread when the queued workfun completes
+void tp_work_done(uv_work_t *req, int status)
+{
+    tp_work_t *p_work = (tp_work_t *) req->data;
+    p_work->notify_cb(p_work->notify_ref);
+    free(p_work);
+}
+
+// Called from the main thread when tp_run_work exits
+JL_DLLEXPORT void jl_tp_queue(tp_workfun_t workfun, void * workfun_arg, tp_notify_cb_t notify_cb, int notify_ref)
+{
+    tp_work_t *p_work = (tp_work_t *)malloc(sizeof(tp_work_t));
+    p_work->req.data = (void*) p_work;
+
+    jl_ptls_t ptls = jl_get_ptls_states();
+
+    p_work->workfun = workfun;
+    p_work->workfun_arg = workfun_arg;
+    p_work->notify_cb = notify_cb;
+    p_work->notify_ref = notify_ref;
+    p_work->current_module = ptls->current_module;
+    p_work->world_age = ptls->world_age;
+
+    // add to queue and return. The julia task waits on a Condition variable
+    // notified by tp_work_done
+    uv_queue_work(jl_io_loop, &p_work->req, tp_run_work, tp_work_done);
+
+    return;
+}
+
+/*
+ * Enable Julia code running in worker threads to forward a f(args)
+ * for execution on the main thread
+ */
+
+
+#include "../deps/srccache/libuv/src/queue.h"
```

Comment:
Can copy it over - https://github.com/JuliaLang/libuv/blob/julia-uv1.9.0/src/queue.h - a simple circular queue implementation, not related to the rest of the libuv stuff.